services/terrain/src/terrain/services/filesystem/metadata.clj

```diff
@@ -377,44 +200,59 @@
 (defn- bulk-add-avus
   "Applies metadata from a list of attributes and filename/values to those files found under
    dest-dir."
-  [cm user dest-dir force? template-id template-attrs attrs csv-filename-values]
+  [user dest-dir force? template-id template-attrs attrs csv-filename-values]
   (let [format-path (partial format-csv-metadata-filename dest-dir)
         paths (map (comp format-path first) csv-filename-values)
+        path-info-map (-> (data-raw/collect-stats user :paths paths :validation-behavior "write") :body json/decode (get "paths"))
         value-lists (map rest csv-filename-values)
         irods-attrs (clojure.set/difference (set attrs) (set (keys template-attrs)))]
-    (validators/all-paths-exist cm paths)
-    (validators/all-paths-writeable cm user paths)
     (if-not force?
-      (validate-batch-add-attrs cm paths irods-attrs))
-  (mapv (partial bulk-add-file-avus cm user dest-dir template-id template-attrs attrs)
-    paths value-lists)))
+      (validate-batch-add-attrs user (map #(get-in path-info-map [% "id"]) paths) irods-attrs))
+  (mapv (partial bulk-add-file-avus user dest-dir template-id template-attrs attrs)
+    path-info-map value-lists)))
+
+(defn- keyword-to-int [kw] (Integer/parseInt (name kw)))
+
+(defn- deparse-csv-line
+  [line]
+  (mapv second (into (sorted-map-by #(< (keyword-to-int %1) (keyword-to-int %2))) line)))
+
+(defn- get-csv
+  [user src separator]
+  (let [path-uuid (data/uuid-for-path user src)
+        chunk-size 1048576 ;; This corresponds to the largest size allowed by the UI's viewer, 1024KB.
+                           ;; It could probably be larger without issue, but this is the expected largest
+                           ;; value for the underlying endpoint in practice. Ideally, it'll never matter
+                           ;; and everything will fit in one page.
+        get-page  (fn [page] (service-response->json (data-raw/read-tabular-chunk user path-uuid separator page chunk-size)))]
+    (loop [page 1 max-pages nil csv []]
+      (let [{max-pages :number-pages new-csv :csv :as res} (get-page page)
+            csv (concat csv (map deparse-csv-line (remove empty? new-csv)))]
+        (if (< page (Integer/parseInt max-pages))
+            (recur (+ page 1) (Integer/parseInt max-pages) csv)
+            csv)))))
```

Comment:
The loop logic won't come up unless the metadata CSV file is bigger than a megabyte, which should hopefully be unlikely.